authenticate.go

```diff
@@ -226,6 +187,127 @@ func (c *OktaClient) cancelCurrentFactor(transaction api.AuthenticationTransacti
+	}
+	return c.handleAuthUserFlow(newTransaction, false)
+}
+
+func (c *OktaClient) handleFactorTypeCode(transaction api.AuthenticationTransaction) (string, error) {
+	code, err := c.prompts.VerifyCode(apiFactorToPublicFactor(transaction.Embedded.Factor))
+	if err != nil {
+		return c.cancelCurrentFactorWithErrorMessage(transaction, "Cancelled")
+	}
+
+	verifyReq := api.FactorVerifyCode{
+		FactorVerify: api.FactorVerify{
+			StateToken: transaction.StateToken,
+		},
+		PassCode: code,
+	}
+	newTransaction, apiError, err := c.sendTransactionRequest(transaction.Links.Next.HREF, &verifyReq)
+	if err != nil {
+		return "", err
+	}
+	if apiError != nil {
+		return c.cancelCurrentFactorWithErrorMessage(transaction, apiError.ErrorSummary)
+	}
+	return c.handleAuthUserFlow(newTransaction, false)
+}
+
+// Logic for handling Okta Verify Push. Given a Authentication Transaction, will make an initial call to send a push notification
+// to user's device then prompts them to accept it. Uses a backoff policy to poll the verify endpoint while waiting on a user to accept.
+// Important to note that if a user times out, the initial verify request will still be on their phone and they'll have to accept/reject it
+// before trying again.
+// TODO: Configurable timeouts
+func (c *OktaClient) handleFactorTypePush(transaction api.AuthenticationTransaction) (string, error) {
+	var newTransaction api.AuthenticationTransaction
+
+	// Sends a request to Okta to push a notification to user's device
+	verifyReq := api.FactorVerifyPush{
+		FactorVerify: api.FactorVerify{
+			StateToken: transaction.StateToken,
+		},
+	}
+	transaction, apiError, err := c.sendTransactionRequest(transaction.Links.Next.HREF, &verifyReq)
+	if err != nil {
+		return c.cancelCurrentFactorWithErrorMessage(transaction, "Cancelled")
+	}
+	if apiError != nil {
+		return c.cancelCurrentFactorWithErrorMessage(transaction, apiError.ErrorSummary)
+	}
+
+	// Prompt user to check their device for an Okta Verify notification
+	c.prompts.VerifyPush()
+
+	// Setup and begin constant backoff policy that retries every 3 seconds with a maximum of 10 attempts (timeout after 30 seconds)
+	backoffPolicy := backoff.WithMaxRetries(backoff.NewConstantBackOff(3*time.Second), 10)
+	operation := func() error {
+		newTransaction, apiError, err = c.sendTransactionRequest(transaction.Links.Next.HREF, &verifyReq)
+		if err != nil {
+			return backoff.Permanent(err)
+		}
+		if apiError != nil {
+			return backoff.Permanent(apiError)
+		}
+		if newTransaction.Status == api.StateSuccess {
+			return nil
+		}
+		if newTransaction.FactorResult == api.FactorResultRejected {
+			fmt.Println("Authentication Request rejected")
+			return backoff.Permanent(&NonFatalAuthError{"Authentication Rejected"})
+		}
+		return &NonFatalAuthError{timeoutErrorMessage}
+	}
+	err = backoff.Retry(operation, backoffPolicy)
+
+	// If error is a NonFatalAuthError (timeout or rejection) then cancel the transaction so we can go through the auth flow again
```

Comment:
Would also just note here about Okta's very weird behavior so that we remember in like 6 months when Okta's finally updated their implementation why we went this route in the first place :)